types/ember-data/index.d.ts

```diff
@@ -0,0 +1,1708 @@
+     * load.
+     */
+    load(): Promise<any>;
+    /**
+     * Reloads this has-many relationship.
+     */
+    reload(): Promise<any>;
+  }
+  /**
+   * An RecordReference is a low level API that allows users and
+   * addon author to perform meta-operations on a record.
+   */
+  class RecordReference {
+    /**
+     * The `id` of the record that this reference refers to.
+     */
+    id(): string;
+    /**
+     * How the reference will be looked up when it is loaded: Currently
+     * this always return `identity` to signifying that a record will be
+     * loaded by the `type` and `id`.
+     */
+    remoteType(): string;
+    /**
+     * This API allows you to provide a reference with new data. The
+     * simplest usage of this API is similar to `store.push`: you provide a
+     * normalized hash of data and the object represented by the reference
+     * will update.
+     */
+    push(UNKNOWN: Promise<any>|{}): void;
+    /**
+     * If the entity referred to by the reference is already loaded, it is
+     * present as `reference.value`. Otherwise the value returned by this function
+     * is `null`.
+     */
+    value(): Model;
+    /**
+     * Triggers a fetch for the backing entity based on its `remoteType`
+     * (see `remoteType` definitions per reference type).
+     */
+    load(): Promise<any>;
+    /**
+     * Reloads the record if it is already loaded. If the record is not
+     * loaded it will load the record via `store.findRecord`
+     */
+    reload(): Promise<any>;
+  }
+  /**
+   * A `ManyArray` is a `MutableArray` that represents the contents of a has-many
+   * relationship.
+   */
+  interface ManyArray<T> extends Ember.MutableArray<T> {}
+  class ManyArray<T> extends Ember.Object.extend(Ember.MutableArray as {}, Ember.Evented) {
+    /**
+     * The loading state of this array
+     */
+    isLoaded: boolean;
+    /**
+     * Metadata associated with the request for async hasMany relationships.
+     */
+    meta: {};
+    /**
+     * Reloads all of the records in the manyArray. If the manyArray
+     * holds a relationship that was originally fetched using a links url
+     * Ember Data will revisit the original links url to repopulate the
+     * relationship.
+     */
+    reload(): PromiseArray<T>;
+    /**
+     * Saves all of the records in the `ManyArray`.
+     */
+    save(): PromiseArray<T>;
+    /**
+     * Create a child record within the owner
+     */
+    createRecord(inputProperties?: {}): T;
+  }
+  /**
+   * A `PromiseArray` is an object that acts like both an `Ember.Array`
+   * and a promise. When the promise is resolved the resulting value
+   * will be set to the `PromiseArray`'s `content` property. This makes
+   * it easy to create data bindings with the `PromiseArray` that will be
+   * updated when the promise resolves.
+   */
+  interface PromiseArray<T> extends Ember.ArrayProxy<T>, Ember.PromiseProxyMixin<PromiseArray<T>> {}
+  class PromiseArray<T> {
```

Comment:
@dfreeman I ran into an interesting situation here: extending a generic class and applying a generic mixin. Here's the code I wanted to write:
```ts
class PromiseArray<T> extends Ember.ArrayProxy<T>
  .extend(Ember.PromiseProxyMixin<PromiseArray<T>>) {}
```

We had a workaround for generic mixins but here we also have a generic type on the left-hand side of `.extend`

Reply:
As soon as the `extends` target changes from being a type to being an expression, any type parameters go out of scope because of that TS issue I linked in the `Enumerable<T>` PR :(

If this were an implementation instead of just a declaration, the best I could imagine this being would be something like:

```ts
interface PromiseArray<T> extends Ember.ArrayProxy<T>, Ember.PromiseProxyMixin<PromiseArray<T>> {}
class PromiseArray<T> extends (
  Ember.ArrayProxy.extend(Ember.PromiseProxyMixin) as typeof Ember.Object
) { ... }
```